File: OrchidCore/src/main/java/com/eden/orchid/impl/compilers/sass/SassImporter.java
package com.eden.orchid.impl.compilers.sass;

import com.eden.common.util.EdenPair;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.resources.resource.OrchidResource;
import com.eden.orchid.utilities.OrchidUtils;
import io.bit3.jsass.importer.Import;
import io.bit3.jsass.importer.Importer;

import javax.inject.Inject;
import java.util.Collection;
import java.util.Collections;

public final class SassImporter implements Importer {

    private final OrchidContext context;

    @Inject
    public SassImporter(OrchidContext context) {
        this.context = context;
    }

    @Override
    public Collection<Import> apply(String url, Import previous) {
        EdenPair<String, String> thisItem = splitPath(url);

        String[] availableFiles = new String[]{
                thisItem.first + "/" + thisItem.second + ".scss",
                thisItem.first + "/" + thisItem.second + ".sass",
                thisItem.first + "/" + "_" + thisItem.second + ".scss",
                thisItem.first + "/" + "_" + thisItem.second + ".sass"
        };

        for (String availableFile : availableFiles) {
            String pathStart = OrchidUtils.normalizePath(splitPath(previous.getAbsoluteUri().getPath()).first);
            String pathEnd = OrchidUtils.normalizePath(availableFile);
            String absoluteUri = OrchidUtils.normalizePath(pathStart + "/" + pathEnd);

            if (absoluteUri.contains("//")) {
                absoluteUri = absoluteUri.replaceAll("//", "/");
            }
            if (absoluteUri.startsWith("/")) {
                absoluteUri = absoluteUri.substring(1);
            }

            OrchidResource importedResource = context.getResourceEntry("assets/css/" + absoluteUri);

            if (importedResource != null) {
                String content = importedResource.getContent();

                if (importedResource.shouldPrecompile()) {
                    content = context.precompile(content, importedResource.getEmbeddedData());
                }

                try {
                    String newURI = "" + OrchidUtils.normalizePath(absoluteUri);
                    Import newImport = new Import(newURI, newURI, content);
                    return Collections.singletonList(newImport);
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        return null;
    }

    private EdenPair<String, String> splitPath(String name) {
        name = name.replaceAll("\\\\\\\\", "/");
        name = name.replaceAll("\\\\", "/");

        if (name.contains("/")) {
            String[] pieces = name.split("/");
            String path = "";

            for (int i = 0; i < pieces.length - 1; i++) {
                path += pieces[i].replaceAll("_", "") + "/";
            }
            String fileName = pieces[pieces.length - 1].replaceAll("_", "");

            return new EdenPair<>(OrchidUtils.normalizePath(path), OrchidUtils.normalizePath(fileName));
        }
        else {
            return new EdenPair<>("", name);
        }
    }
}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/compilers/sass/SassImporter.kt
package com.eden.orchid.impl.compilers.sass

import com.caseyjbrooks.clog.Clog
import com.eden.common.util.EdenPair
import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.resources.resource.OrchidResource
import com.eden.orchid.api.theme.Theme
import com.eden.orchid.utilities.OrchidUtils
import io.bit3.jsass.Compiler
import io.bit3.jsass.Sass2ScssOptions
import io.bit3.jsass.importer.Import
import io.bit3.jsass.importer.Importer
import org.apache.commons.io.FilenameUtils
import javax.inject.Inject

class SassImporter(
    private val context: OrchidContext,
    sourceResource: OrchidResource?,
    private val sourceTheme: Theme?,
    private val sourceData: Map<String, Any?>
) : Importer {

    private val ignoredPreviousValues = arrayOf("stdin")
    private val importContext = mutableListOf(sourceResource)

    override fun apply(url: String, previous: Import): Collection<Import>? {
        val (isAbsolute, cleanedCurrentImportName) = cleanInputName(url)
        val preferredInputExtension = getPreferredInputExtension(url)
        val currentDirectory = getCurrentDirectory(url, previous, isAbsolute)

        val availableFiles = when (preferredInputExtension) {
            SassCompiler.CompilerSyntax.SCSS -> cleanResourcePaths(
                getScssPatterns(
                    currentDirectory,
                    cleanedCurrentImportName
                )
            )
            SassCompiler.CompilerSyntax.SASS -> cleanResourcePaths(
                getSassPatterns(
                    currentDirectory,
                    cleanedCurrentImportName
                )
            )
            SassCompiler.CompilerSyntax.UNSPECIFIED -> cleanResourcePaths(
                getAnyPatterns(
                    currentDirectory,
                    cleanedCurrentImportName
                )
            )
        }

        for (availableFile in availableFiles) {
            val importedResource = if(sourceTheme != null) context.getResourceEntry(sourceTheme, availableFile, null) else context.getResourceEntry(availableFile, null)

            if (importedResource != null) {
                importContext.add(importedResource)

                var content = importedResource.content
                if (importedResource.shouldPrecompile()) {
                    content = context.compileWithContextData(
                        importedResource,
                        importedResource.precompilerExtension,
                        content,
                        sourceData
                    )
                }

                try {
                    val baseUri = "" + OrchidUtils.normalizePath(
                        FilenameUtils.removeExtension("$currentDirectory/$cleanedCurrentImportName")
                    )
                    val relativeUri = "" + OrchidUtils.normalizePath(
                        FilenameUtils.removeExtension("$currentDirectory/$cleanedCurrentImportName").split("/").last() + preferredInputExtension.ext
                    )

                    if (importedResource.reference.extension == "sass") {
                        content = convertSassToScss(content)
                    }

                    val newImport = Import(relativeUri, baseUri, content)

                    return listOf(newImport)
                } catch (e: Exception) {
                    e.printStackTrace()
                }

            }
        }

        return null
    }

    private fun splitPath(originalName: String): EdenPair<String, String> {
        val name = cleanPath(originalName)

        if (name.contains("/")) {
            val pieces = name.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            var path = ""

            for (i in 0 until pieces.size - 1) {
                path += pieces[i].replace("_".toRegex(), "") + "/"
            }
            val fileName = pieces[pieces.size - 1].replace("_".toRegex(), "")

            return EdenPair(
                OrchidUtils.normalizePath(path),
                OrchidUtils.normalizePath(FilenameUtils.removeExtension(fileName))
            )
        } else {
            return EdenPair("", FilenameUtils.removeExtension(originalName))
        }
    }

    private fun cleanPath(originalName: String): String {
        var name = originalName
        name = name.replace("\\\\\\\\".toRegex(), "/")
        name = name.replace("\\\\".toRegex(), "/")

        return name
    }

    private fun getScssPatterns(baseDir: String, importPath: String): Array<String> {
        return arrayOf(
            "$baseDir/$importPath.scss",
            "$baseDir/_$importPath.scss"
        )
    }

    private fun getSassPatterns(baseDir: String, importPath: String): Array<String> {
        return arrayOf(
            "$baseDir/$importPath.sass",
            "$baseDir/_$importPath.sass"
        )
    }

    private fun getAnyPatterns(baseDir: String, importPath: String): Array<String> {
        return arrayOf(
            *getScssPatterns(baseDir, importPath),
            *getSassPatterns(baseDir, importPath)
        )
    }

// helpers
//----------------------------------------------------------------------------------------------------------------------

    private fun cleanInputName(import: String): Pair<Boolean, String> {
        return Pair(
            import.trim().startsWith("/"),
            splitPath(import).second
        )
    }

    private fun getPreferredInputExtension(input: String): SassCompiler.CompilerSyntax {
        return when (FilenameUtils.getExtension(input)) {
            SassCompiler.CompilerSyntax.SCSS.ext -> SassCompiler.CompilerSyntax.SCSS
            SassCompiler.CompilerSyntax.SASS.ext -> SassCompiler.CompilerSyntax.SASS
            else                                 -> SassCompiler.CompilerSyntax.UNSPECIFIED
        }
    }

    private fun getCurrentDirectory(import: String, previous: Import, isAbsolute: Boolean): String {
        val baseDirectory = if (ignoredPreviousValues.any { it == previous.absoluteUri.normalize().toString() }) {
            "assets/css"
        } else {
            previous.absoluteUri.normalize().toString().split("/").dropLast(1).joinToString("/")
        }

        val importDirectory = splitPath(import).first

        return if(isAbsolute) {
            importDirectory
        }
        else {
            if(importDirectory.isBlank()) {
                OrchidUtils.normalizePath(baseDirectory)
            }
            else {
                OrchidUtils.normalizePath("$baseDirectory/$importDirectory")
            }
        }
    }

    private fun cleanResourcePaths(inputs: Array<String>): List<String> {
        return inputs.map { OrchidUtils.normalizePath(it) }
    }

    private fun convertSassToScss(input: String): String {
        return Compiler.sass2scss(input, Sass2ScssOptions.CONVERT_COMMENT)
    }
}
